### PR TITLE
Feature/dashboard redesign screenshots 96111874

### PR DESCRIPTION
**Pull request by crsven**
* Screenshots are taken on the client after 3 seconds of inactivity, shrunk by half, and socketed to the server
* Adds AWS storage for screenshots
* Actually added some specs!

screenshot 2015-06-09 16 14 52

### CODE REVIEW

File: lib/services/sockets.coffee [hunk 1]
```diff
@@ -80,6 +88,39 @@ class Sockets
...
+
+        socket.on 'board.uploadImage', (message) ->
+          Board.findSession socket.id, (err, board, user) ->
+            if board && user
+              redisClient.get "board_#{board.id}_collaborator_chosen", (err, chosenCollaborator) ->
+                if user.id == chosenCollaborator
+                  matches = message.image.match(/^data:([A-Za-z-+\/]+);base64,(.+)$/)
+                  imageData = {}
+
+                  if matches.length != 3
```
Comment (andrewhao) [inline lib/services/sockets.coffee]: What is the significance of `3`?

File: assets/js/handler.coffee [hunk 2]
```diff
@@ -131,6 +132,30 @@ class boardroom.Handler
...
+  onScreenshot: (message) =>
+    if @user.get('userId').toString() == message.userId.toString()
+      boardSize = @board.getBoardSize()
+      height = boardSize.y + 44
+      width = boardSize.x + 80
+      html2canvas($('.board')[0], {
+        height: height,
+        width: width,
+        logging: true,
+      }).then (canvas) =>
```
Comment (andrewhao) [inline assets/js/handler.coffee]: Is there any value in logging errors here? Perhaps sending to Airbrake?

File: assets/js/handler.coffee [hunk 3]
```diff
@@ -131,6 +132,30 @@ class boardroom.Handler
...
+        image = context.getImageData(0, 0, width, height)
+        canvas.height = boardSize.y / 2
+        canvas.width = boardSize.x / 2
+
+        newCanvas = $("<canvas>")
+          .attr("width", width)
+          .attr("height", height)[0]
+        newCanvas.getContext("2d").putImageData(image, 0, 0)
+
+        context.scale(0.5, 0.5)
```
Comment (andrewhao) [inline assets/js/handler.coffee]: Why scale? Bandwidth optimization? How large are the images?

Comment (andrewhao) [inline assets/js/handler.coffee]: (Also: are the images transparent PNGs?)

**Comment by andrewhao** [comment]
:+1:. See minor nits and questions, but overall lgtm.

File: assets/js/handler.coffee (hunk 3 above)
Comment (gmas) [inline assets/js/handler.coffee]: @andrewhao yes, the PNGs have an alpha channel. 
the size of the images depends on how big the board is. we scale them to save on bandwidth/space

File: lib/services/sockets.coffee (hunk 1 above)
Comment (gmas) [inline lib/services/sockets.coffee]: the regex has to match all 2 groups:
- the whole regex
- "data:" group (image/png)
- the encoded image data

File: assets/js/handler.coffee (hunk 2 above)
Comment (gmas) [inline assets/js/handler.coffee]: sure, we can add it

File: lib/services/sockets.coffee (hunk 1 above)
Comment (andrewhao) [inline lib/services/sockets.coffee]: Ah! gotcha.

File: lib/services/sockets.coffee [hunk 4]
```diff
@@ -80,9 +89,48 @@ class Sockets
...
+                  uploader.upload imageData.data, fileName, (err, imageUrl) ->
+                    if !err
+                      board.updateAttributes { screenshotUrl: imageUrl.Location }, ->
+
         socket.on 'log', ({user, board, level, msg}) =>
           logger.logClient user, board, level, msg
 
+        socket.on '_error', (error) ->
+          console.log "Airbraking _error: ", error
+          airbrakeClient = airbrake.createClient process.env.AIRBRAKE_API_KEY
```
Comment (andrewhao) [inline lib/services/sockets.coffee]: Maybe we can use the airbrakeClient in the `Config` object?

**Pull request merged**